### PR TITLE
use php-di decorator

### PR DESCRIPTION
**Pull request by gianarb**
Use php-di decorator feature to improve the attach listeners flow

At the moment not work and I don't understand why.. :+1:

### CODE REVIEW

File: config/di.php [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
+        $eventManager->attach('ERROR_DISPATCH', function (\GianArb\Penny\Event\HttpFlowEvent $event) use ($template) {
+            $e = $event->getException();
+
+            $response = $event->getResponse()->withStatus($e->getCode());
+            $event->setResponse($response);
+
+            $response->getBody()->write($template->render("error/40x", [
+                'title' => $e->getMessage(),
+                'exception' => $e,
+            ]));
```
Comment (samsonasik) [inline config/di.php]: missing return `$eventManager` ?

Comment (mnapoli) [inline config/di.php]: sounds like it! That's a common mistake unfortunately :/ (I did it too) I'm still looking for a solution to help prevent that. The thing is that a decorator could return `null`, so throwing an exception if a decorator returns `null` isn't good… Ideas welcome!

**Comment by samsonasik** [comment]
I've created PR #22 for that ;)

**Comment by samsonasik** [comment]
@gianarb seems still not work, the issue is as the eventmanager registered as service, it not automatically called ?

**Comment by gianarb** [comment]
Thanks @samsonasik and @mnapoli  but the problem persist..
decorator function seems not to be call

@samsonasik the event_manager service is defined by penny framework.. I'm trying to decorate it with the listeners

**Comment by samsonasik** [comment]
@gianarb if I debug: 

``` php
var_dump(
    $app->getContainer()->get('event_manager')
);
```

in `public/index.php`, it seems call old `event_manager` service instead of new one in `config/di.php`

**Comment by samsonasik** [comment]
@gianarb probably related with `Loader::load()` called first before `$container->set('event_manager')` ?

File: config/di.php (hunk 1 above)
Comment (gianarb) [inline config/di.php]: @mnapoli can you return $prevent if the callback haven't return value?

**Comment by gianarb** [comment]
@samsonasik I have the same thought..

File: config/di.php (hunk 1 above)
Comment (mnapoli) [inline config/di.php]: @gianarb returning nothing and returning `null` is the same thing in PHP. And sometimes you want to return null so that a service or parameter is set to `null`. So doing what you suggest would break that.

**Comment by mnapoli** [comment]
`$container->set()` adds the definition _at the end of the definitions list_. In other words, it will override any previous definition.

So `$container->set('event_manager', ...)` will indeed override all previous definitions (e.g. `decorate()` definitions).

In general I really advise against dynamically setting services in the container with `->set()`. Most services should be defined statically in container definitions. One of the reason for this is performances (same reason Symfony does that: Symfony's container isn't even mutable at all). Another reason (or rather a fortunate consequence) is that it forces applications to be statically wired/configured so you have much more control on how it's built, on service overriding, etc.

So TL/DR: favor array definitions over `$container->set()`

**Comment by gianarb** [comment]
it wooorksss!!!!! :)

**Comment by samsonasik** [comment]
I think PageNotFound listener can be changed with ClientError listener for naming ;). So, for 50x, we can create ServerError listener

**Comment by gianarb** [comment]
ClientError or DispatchError? :)

**Comment by samsonasik** [comment]
DispatchError is ok I think (y)

Warm regards,

Abdul Malik Ikhsan

Pada 24 Sep 2015, pukul 14.04, Gianluca Arbezzano notifications@github.com menulis:

**Comment by fntlnz** [comment]
LGTM :+1:

File: test/AppTest.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            ->getMock();
+        $spy->expects($this->once())
+            ->method('onError')
+            ->with($this->anything());
+
+        $response = $this->getMock(Response::class);
+        $request = (new Request())
+            ->withUri(new Uri('/pnf'))
+            ->withMethod('GET');
+
```
Comment (samsonasik) [inline test/AppTest.php:27]: I think it can be simplified with prophecy to make consistent with controller tests.

``` php
use Prophecy\Argument;

$spy = $this->prophesize(DispatcherExceptionListener::class);
$spy->onError(Argument::any())->shouldBeCalled();

$response = $this->prophesize(Response::class);
// ..
$container->set(DispatcherExceptionListener::class, $spy->reveal());
// ...
$app->run($request, $response->reveal());
```

Comment (gianarb) [inline test/AppTest.php:27]: I wait your PR :D I don't know prophecy but I'm very happy to learn it!

Comment (samsonasik) [inline test/AppTest.php:27]: done ;) https://github.com/pennyphp/penny-classic-app/pull/25
